File: day25/day25.cpp
#include <algorithm>
#include <fstream>
#include <iostream>
#include <unordered_map>
#include <vector>

#include <fmt/format.h>

using namespace std;

struct Instruction
{
	int curstate;
	int curval;
	int newval;
	int move;
	int nxtstate;
};

struct Machine
{
	int state;
	size_t checksum;
	unordered_map<int, int> tape;
	vector<Instruction> instructions;

	size_t run() noexcept
	{
		int p = 0;
		for (size_t j = 0; j < checksum; j++)
		{
			auto value = tape[p];
			auto it = find_if(instructions.begin(),
					  instructions.end(),
					  [this, value](const auto& i) {
						  return i.curstate == state && i.curval == value;
					  });
			if (it != instructions.end())
			{
				tape[p] = it->newval;
				p += it->move;
				state = it->nxtstate;
			}
		}
		return count_if(
			tape.begin(),
			tape.end(),
			[](const auto& p) {
				return p.second == 1;
			});
	}
};

istream& operator>>(istream& in, Machine& m)
{
	Instruction i;
	string line;
	while (getline(in, line))
	{
		if (line.rfind("Begin in state ", 0) == 0)
		{
			m.state = line[15];
		}
		else if (line.rfind("Perform a diagnostic checksum after ", 0) == 0)
		{
			m.checksum = stoull(line.substr(36));
		}
		else if (line.rfind("In state ", 0) == 0)
		{
			i.curstate = line[9];
		}
		else if (line.rfind("  If the current value is ", 0) == 0)
		{
			i.curval = stoi(line.substr(26));
		}
		else if (line.rfind("    - Write the value ", 0) == 0)
		{
			i.newval = stoi(line.substr(22));
		}
		else if (line.rfind("    - Move one slot to the ", 0) == 0)
		{
			i.move = line.substr(27) == "right." ? 1 : -1;
		}
		else if (line.rfind("    - Continue with state ", 0) == 0)
		{
			i.nxtstate = line[26];
			m.instructions.emplace_back(move(i));
		}
	}
	return in;
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		cerr << fmt::format("Usage: {} <filename>", argv[0]) << endl;
		return 1;
	}

	ifstream input(argv[1]);
	if (!input)
	{
		cerr << fmt::format("Cannot open {}", argv[1]) << endl;
		return 1;
	}
	Machine m;
	input >> m;
	input.close();

	cout << fmt::format("Part1: {}", m.run()) << endl;
	return 0;
}

File: day25/day25.py
#!/usr/bin/env python3

def parse(txt):
    begin = None
    steps = None
    states = {}
    state, current, new, move, nxt = None, None, None, None, None
    for line in txt.splitlines():
        if line.startswith("Begin in state "):
            begin = line[15]
        elif line.startswith("Perform a diagnostic checksum after "):
            steps = int(line[36:].split(" ")[0])
        elif line.startswith("In state "):
            state = line[9]
        elif line.startswith("  If the current value is "):
            current = int(line[26:-1])
        elif line.startswith("    - Write the value "):
            new = int(line[22:-1])
        elif line.startswith("    - Move one slot to the "):
            move = (line[27:] == "right.") and 1 or -1
        elif line.startswith("    - Continue with state "):
            nxt = line[26]
            states[state, current] = (new, move, nxt)

    return begin, steps, states

def run(txt):
    curstate, steps, states = parse(txt)
    v = {}
    p = 0
    for i in range(steps):
        curval = v.get(p, 0)
        newval, move, nxt = states[curstate, curval]
        v[p] = newval
        p += move
        curstate = nxt

    return sum(1 for i in v.values() if i == 1)

import sys
if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read().strip()
    except:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)

    print(run(txt))
    # TODO: implement the code
    # print("Part1:", tls)
    # print("Part2:", ssl)
